cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -190,12 +190,13 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 	if err != nil {
 		return nil, errors.Wrap(err, "error syncing endpoints with etc")
 	}
+	klog.V(1).Infof("update etcd endpoints: %s", strings.Join(etcdClient.Endpoints, ","))
 
 	return etcdClient, nil
 }
 
 // Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
-func (c Client) Sync() error {
+func (c *Client) Sync() error {
 	cli, err := clientv3.New(clientv3.Config{
```

Comment:
having a single method that accepts a pointer and the rest accepting non-pointers seems a bit odd.
we might want to make all of them accept pointer... see GetVersion(),  GetClusterVersions() etc.

Reply:
you mean that Sync() return the real etcd endpoints? such as:

```go
func (c Client) Sync() ([]string, error) {}
```

Reply:
the methods for `Client` have different signatures:
`func (c *Client) Sync()`
`func (c Client) AddMember(name string, peerAddrs string) ([]Member, error) {`
...

`Sync()` uses a pointer, but the rest of the methods use non-pointers.

Reply:
Sync() only return error now, but the rest of the methods return multi values.
Can change this `Sync()` method return endpoints and error, like other methods?
The `NewFromCluster` invoke `Sync()` like below:

```go
endpoints, err = etcdClient.Sync()
if err != nil {
	return nil, errors.Wrap(err, "error syncing endpoints with etc")
}
etcdClient.Endpoints = endpoints
```
In this case, all methods use non-pointers, and only change `Sync()` methods, but test case maybe need to update.

The other ways, the methods of `Client` all change non-pointers to pointer, the way need to change more codes.

I am not sure if i understand what you mean？If wrong, can you explain your comments again. Thanks.

Reply:
@pytimer my point is this:
```
func (c *Client) Sync(...) ...
func (c Client) AddMember(...) ...
...
```
^ see how `c` is a pointer in `Sync` but not a pointer in the rest of methods for the `Client` type.
we need to make all the methods consistent.

Reply:
@pytimer I'm ok with the solution you are proposing:
```
func (c Client) Sync() ([]string, error) {}
```

@neolit123 IMO the solution above keeps consistency in the signature and address the issue.
Eventually we will reconsider changing all the methods in another PR

Reply:
/approve
as per @fabriziopandini 's comment.
thanks @pytimer

Reply:
/hold

Reply:
@pytimer please fix as discussed above

Reply:
@fabriziopandini @neolit123 

Fix like:

```go
func (c Client) Sync() ([]string, error) {}
```

Is it ok?

Reply:
@pytimer @fabriziopandini @neolit123 I am not OK with `func (c Client) Sync() ([]string, error)`. This is no longer a `Sync` method. It's not even a `GetEndpoints` method (because we have `Endpoints` member in `Client`). Furthermore, this introduces a code smell (inappropriate intimacy), where `NewFromCluster` now updates field inside `Client`.

This is a lot worse solution to the problem it tries to solve. I am much more in favor of making all methods `func (c *Client)`, than have this. I am also in favor of having only `Sync()` take a pointer to `Client`.

Reply:
@pytimer 
OK, i missed the part that we want to modify the existing Client object and not a copy.
please bring back `(c *Client)` .